src/mixins/yield-erc/ERC20Vault.sol

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity 0.8.10;
+
+import {ERC20} from "../../tokens/ERC20.sol";
+import {FixedPointMathLib} from "../../utils/FixedPointMathLib.sol";
+import {SafeTransferLib} from "../../utils/SafeTransferLib.sol";
+
+/// @title Yield Bearing Vault
+/// @author joeysantoro, Transmissions11 and JetJadeja
+contract ERC20Vault is ERC20 {
+    using SafeTransferLib for ERC20;
+    using FixedPointMathLib for uint256;
+
+    /*///////////////////////////////////////////////////////////////
+                                IMMUTABLES
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice The underlying token the Vault accepts.
+    ERC20 public immutable underlying;
+
+    /// @notice The base unit of the underlying token and hence vault.
+    /// @dev Equal to 10 ** decimals. Used for fixed point arithmetic.
+    uint256 public immutable baseUnit;
+
+    /// @notice Creates a new Vault that accepts a specific underlying token.
+    /// @param _underlying The ERC20 compliant token the Vault should accept.
+    /// @param _name The name for the vault token.
+    /// @param _symbol The symbol for the vault token.
+
+    constructor(
+        ERC20 _underlying,
+        string memory _name,
+        string memory _symbol
+    )
+        ERC20(
+            _name,
+            _symbol,
+            _underlying.decimals()
+        )
+    {
+        underlying = _underlying;
+
+        baseUnit = 10**decimals;
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                        DEPOSIT/WITHDRAWAL LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Emitted after a successful deposit.
+    /// @param user The address that deposited into the Vault.
+    /// @param underlyingAmount The amount of underlying tokens that were deposited.
+    event Deposit(address indexed user, uint256 underlyingAmount);
+
+    /// @notice Emitted after a successful withdrawal.
+    /// @param owner The address that withdrew from the Vault.
+    /// @param to The destination for withdrawn tokens.
+    /// @param underlyingAmount The amount of underlying tokens that were withdrawn.
+    event Withdraw(address indexed owner, address indexed to, uint256 underlyingAmount);
+
+    /// @notice Deposit a specific amount of underlying tokens.
+    /// @param to The address to receive shares corresponding to the deposit
+    /// @param underlyingAmount The amount of the underlying token to deposit.
+    function deposit(address to, uint256 underlyingAmount) external virtual returns (uint256 shares) {
+        shares = underlyingAmount.fdiv(exchangeRate(), baseUnit);
+        // Determine the equivalent amount of shares and mint them.
+        _mint(to, shares);
```

Comment:
Given this line 

https://github.com/Rari-Capital/solmate/blob/ab1d04a2ba35cd63ebe90d3704ff0a010a576246/src/tokens/ERC20.sol#L181

a deposit to address 0x0 emits an event that looks like a burn to subgraph and other trackers when the supply actually hasnt changed.

See item 1 in audit for yearn vaults, thought it was similar.

https://github.com/yearn/yearn-security/blob/master/audits/20210719_ToB_yearn_vaultsv2/ToB_-_Yearn_Vault_v_2_Smart_Contracts_Audit_Report.pdf

Reply:
argh good point, would hate to impl a to != 0 check but seems like this might be a valid case

Reply:
i think we might even want to impl this at the ERC20 level? hmmmmmmm

Reply:
well wait this is already possible just by transferring shares to address(0). and we're not the only erc20 that supports doing that (see UNIv2ERC20)

Reply:
Not sure if we need to address this, will let @transmissions11 make the final call

Reply:
yeah, makes sense to leave as is , given that is understood by users of the lib.

Reply:
why _mint going before transfer? It may cause re-entrant attack. So _mint should be after tranferring